CRM/Contact/BAO/Group.php

```diff
@@ -1016,15 +1016,37 @@ public static function getGroupList(&$params) {
         $values[$object->id]['created_by'] = "<a href='{$contactUrl}'>{$object->created_by}</a>";
       }
 
-      // get group contact count using Contact.GetCount API
-      // CRM-20226 This has been added here in order to address a specific bug. However, a prior
-      // decision was to refresh group counts less aggressively, offering instead a button to
-      // refresh them to give users a better experience by loading pages quicker.
-      // For some sites this can be crazy slow even though only 25 sites resolve. Even for sites
-      // with relatively few smart groups it is not a good user experience.
-      // Adding comments here as I'm not going to tackle a fix this time around but want
-      // to warn people off making it worse.
-      $values[$object->id]['count'] = civicrm_api3('Contact', 'getcount', array('group' => $object->id));
+      // By default, we try to get a count of the contacts in each group
+      // to display to the user on the Manage Group page.
+      $skip_getcount = FALSE;
+      // If it's a smart group AND the contacts aren't cached, don't try to
+      // generate a count, it will take forever.
+      if ($object->saved_search_id) {
+        if (is_null($object->cache_date)) {
+          $skip_getcount = TRUE;
+        }
+        else {
+          // We have a cache_date - see if it is expired.
+          $params['name'] = 'smartGroupCacheTimeout';
+          $timeout = civicrm_api3('Setting', 'getvalue', $params);
+          $cache_expires = date('Y-m-d H:i:s', time() - (intval($timeout) * 60));
+          if ($cache_expires > $object->cache_date) {
+            $skip_getcount = TRUE;
+          }
+        }
+      }
+      if ($object->children) {
+        // If there are children, any of the children could be expired
+        // smart groups.
+        $skip_getcount = TRUE;
```

Comment:
I am not sure if we simply need to place ```unknown``` text instead of count for any parent group, some might not go with this tweak. @eileenmcnaughton what if we wrap up this change under any special setting. 

Also need input from @lcdservices @colemanw

Reply:
Oh. Reverting CRM-20226 was an unintended consequence of adding the check for children :(. 

That means parent groups will always display an unknown count even if all smart groups have been refreshed.

One solution would be to write a function that takes a group id and returns TRUE if it is a smart group with a cache that has expired or FALSE otherwise. Then, if a group has children, we iterate over all children and check each one against this function. If a group or any children return TRUE, we set to unknown. Otherwise, we allow the count to happen.

Reply:
Ha, function seems to exist: CRM_Contact_BAO_GroupContactCache:shouldGroupBeRefreshed(). Although it handles parent groups differently then I expected. It seems that parent groups are cached like smart groups?

If that's the case, (and if parent groups are properly refreshed and cached when a user click the link to "Update smart group counts") - then we can simplify everything by simply checking the group ID against this function to determine whether to display "Unknown" or not.

Reply:
Take three. This time I try to preserve what was added by CRM-20226 and I've added a few more tests to ensure we get both unknown and a count under the right conditions.

To be perfect, we could probably use another test of a parent group with a smart group child but I'm running out of steam...